### DISCUSSION TITLE
Add `managed` constraint, inverse of `unmanaged` constraint

### DISCUSSION THREAD

**Discussion by jthorborg** [dotnet/csharplang#6143] [Language Ideas]
Since constraints are not a part of the signature, you cannot overload methods on constraints. However... You can achieve this using static extensions, and actually implement a sort of SFINAE.

This is interesting to dispatch calls to particular constrained versions when offering generic APIs, eg. have a fast path for `unmanaged` types. It almost works, except that there's not a perfect inverse constraint, you only have `class` and `struct`. `struct` intersects `unmanaged` so overload resolution becomes ambiguous.

Since this is something the compiler already analyses and understands (and all of this happens in the front end anyway), it seems like it is theoretically doable? Here's sketch showcasing the problem today:

```c#
public struct SomeAPI { }

struct ManagedStruct
{
    string ThisCausesItToBeManagedButStillAValueType;   
}

public static class SomeAPIUnmanagedExtensions
{
    // hurray, we can take pointers to argument and have a fast path
    public static void AnAPI<T>(this SomeAPI apiObject, T argument)
        where T : unmanaged
    {
    }
}


public static class SomeAPIManagedExtensions
{
    // slow but flexible path.
    public static void AnAPI<T>(this SomeAPI apiObject, T argument)
        // Ideally, we could have "managed" constraint here.
        where T : class
    {
    }
}

public static class SomeAPIStructExtensions
{
    /*public static void AnAPI<T>(this SomeAPI apiObject, T argument)
        where T : struct
    {
        // specialization for structs. doesn't work unfortunately, see below.
    } */
}


public class Program 
{
    public void Usage() 
    {
        SomeAPI instance;
        
        instance.AnAPI(3.14f);
        instance.AnAPI("hello, world");
        
        // Will not match either class or unmanaged constraint
        // Cannot uncomment SomeAPIStructExtensions, as 
        // unmanaged <> struct overloads then become ambiguous
        instance.AnAPI(new ManagedStruct());
    }
}

```

**Comment by jnm2** [comment]
The compiler doesn't allow overloading on `struct` vs `class` constraints either. What it allows is overloading on differing parameter types, such as `T` and `Nullable<T>`.

**Comment by huoyaoyuan** [comment]
Generally, we use constraints to allow additional functionality under the constraint. Unmanaged constraint allows pinning etc.
